### ISSUE TITLE
Color word change can break a lot of cards

### ISSUE THREAD

**Issue by tool4ever** [Card-Forge/forge#2512]
Especially devastating if you use _Swirl the Mists_:
I noticed when _Quicksilver Elemental_ stopped working because it was looking for "RemembewhiteLKI". :/

**Comment by Hanmac** [comment]
we probably need a better way to teach it what are color words are and what not and what it should replace and what not.

like inserting a \b into the Regex to search for Word Bounds? that should find "+Red" while ignoring "Remembered"

**Comment by tool4ever** [comment]
I'll try that

Also the negative lookbehind to avoid matching inside <strike> should only be used when descriptive, otherwise the second color in matrix syntax might fail to match, e.g. "White->Red" from _Sunglasses of Urza_

**Comment by Hanmac** [comment]
yeah, separate the descriptive changes 

i thought that color only appears as Properties so that we can use `\b(non)?(red|blue|white|black|green)(Source)?` 
Should we also clean up the "non" negation?

as for the descriptive changes yeah these are a bit tricky ... one of the things why i want to refactor Keywords more is to make them be able to show such changes. currently Keywords does bypass the change and they are created new

I don't remember where but i think there was once a piece of code that was trying to revert the card text changes from back to the original (like if you change Red to White, and it finds White in the Description it tries to get Red instead? But that code seems to be removed

**Comment by tool4ever** [comment]
right, gotta keep the negations intact as ruled on cards like _Trait Doctoring_:
>The type-changing effect can change part of a word such as “nonblack” or “swampwalk” if the part of the word is being used to refer to a color or basic land type.

Maybe we should convert `non` -> `!` inside `hasProperty()` calls to get the front word boundary
- then scripts can still use it for closer text mirroring
- we could clean up lots of doubled "non..." properties

Still need to think if we also need end boundary, but might be tricky to handle cases such as landwalk without exception....?

**Comment by Hanmac** [comment]
stuff like landwalk is going to be refactored by me later in #2507 into `K:Landwalk:Swamp` anyway so you don't need to care about that for now

one of the tricky parts for description is stuff like "colo**red**" thats why we try to avoid it for the Properties

maybe we should use some kind of Key in the Description like `non<color>black</color>` so when the Descriptive Changes are done, it knows what kind to replace?

**Comment by zzo38** [comment]
Text-changing effects ought to be AST-based, i.e. there is a node whose value is the atom `blue`, `green`, etc, and it replaces all such tokens in the AST of the object's text (where "text" is one of its characteristics, effectively being the AST), and therefore would affect the behaviour. In the case of words such as "nonblack", this is effectively like a node with AST `not(black)`, so it would still be affected, since one of its child nodes is a color word.

For changing printed text, perhaps things like `non<color>black</color>` would help, and then the display might do something such as cross out the word "black" and write "red" in a another text style.

**Comment by Hanmac** [comment]
maybe we could add some kind of `{cw:red}` or something like that or maybe `{crt:dragon}` for Descriptive Parameters
or do the negative like `{name:Pyrexian Dragon Engine}`

so when something wants to alter creature types inside description, it knows that it should touch `Dragons you control` while ignoring the part of `named Pyrexian Dragon Engine`

This is just about the Description part, the normal Params should already be ignored by it, i think

**Comment by tool4ever** [comment]
So something like _Dragon's Approach_ works correctly (for non-descriptive at least) but when I used _Artificial Evolution_ on the token from _Deathpact Angel_ it breaks currently because `Card.namedDeathpact Angel` will match with the new word boundary approach :/

Shouldn't be too hard to fix surgically though, I could either
1. just edit the few scripts with evil type names to use _ instead of whitespace (already supported by the parsing)
2. try to skip both in `getReplacedText` when "named" is included

**Comment by tool4ever** [comment]
Hmn I'm not sure if that wouldn't mess up the other languages auto feature of CardTranslation too much?

But if plural types should never occur in non-descriptive params we could at least tweak it so it doesn't waste time attempting to regex on it?

**Comment by Hanmac** [comment]
@tool4ever i don't know about the auto feature of CardTranslation for ingame Description of CardTraits

The goal is that the Card Trait says `{pt:Dragon} you control` which then the Description would into `Dragons you control`
this should also be that what the CardTranslation is seeing?

now when the CardTextChange wants to alter the Description, it changes to `{pl:Angel} you control` into `Angels you control`
that would be problematic for CardTranslation anyway, right?

the only goal there is that the TextChange doesn't need to take care about plural types anymore, because Plural would be done after TextChanges
